pkg/upgrade/machine_creator.go

```diff
@@ -0,0 +1,283 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterapiv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type podGetter interface {
+	Get(string, metav1.GetOptions) (*v1.Pod, error)
+}
+
+type nodeLister interface {
+	List(options metav1.ListOptions) (*v1.NodeList, error)
+}
+
+type machineGetter interface {
+	Get(string, metav1.GetOptions) (*clusterapiv1alpha1.Machine, error)
+}
+
+type machineCreator interface {
+	Create(*clusterapiv1alpha1.Machine) (*clusterapiv1alpha1.Machine, error)
+}
+
+// MachineOptions are the values that can change on a machine upgrade
+type MachineOptions struct {
+	ImageID string
+	ImageField string
+	DesiredVersion semver.Version
+}
+
+// MachineCreator is responsible for creating a new machine.
+// It can optionally wait for the associated node to become ready.
+type MachineCreator struct {
+	WaitForProviderID bool
+	WaitForMatchingNode bool
+	WaitForNodeReady bool
+	MachineOptions MachineOptions
+
+	podGetter podGetter
+	nodeLister nodeLister
+	machineGetter machineGetter
+	machineCreator machineCreator
+	log logr.Logger
+}
+
+// NewMachineCreator takes a list of option functions to configure the MachineCreator.
+func NewMachineCreator(options ...MachineCreatorOption) *MachineCreator {
+	creator := &MachineCreator{}
+	for _, fn := range options {
+		fn(creator)
+	}
+	return creator
+}
+
+// NewMachine is the main interface to MachineCreator.
+// It creates a machine object on the management cluster and optionally waits for the backing node to become ready.
+func (n *MachineCreator) NewMachine(source *clusterapiv1alpha1.Machine) (*clusterapiv1alpha1.Machine, *v1.Node, error) {
+	newMachine := source.DeepCopy()
+
+	// have to clear this out so we can create a new machine
+	newMachine.ResourceVersion = ""
+
+	// have to clear this out so the new machine can get its own provider id set
+	newMachine.Spec.ProviderID = nil
+
+	// assume the original name is controlplane-<index> or controlplane-<index>-<timestamp>
+	// let's set the new name to controlplane-<index>-<timestamp>
+	nameParts := strings.Split(source.Name, "-")
+	if len(nameParts) < 2 {
+		return nil, nil, errors.Errorf("machine name %q does not match expected format <name>-<index> or <name>-<index>-<timestamp>", source.Name)
+	}
+	newMachine.Name = fmt.Sprintf("%s-%s-%d", nameParts[0], nameParts[1], time.Now().Unix())
+
+	if n.MachineOptions.ImageField != "" && n.MachineOptions.ImageID != "" {
+		if err := updateMachineSpecImage(&newMachine.Spec, n.MachineOptions.ImageField, n.MachineOptions.ImageID); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	newMachine.Spec.Versions.ControlPlane = n.MachineOptions.DesiredVersion.String()
+	newMachine.Spec.Versions.Kubelet = n.MachineOptions.DesiredVersion.String()
+
+	n.log.Info("Creating new machine", "name", newMachine.GetName())
```

Comment:
Personal preference, but I tend to access the fields directly (`.Name`) vs calling the accessor functions (`GetName()`, etc.). Do you have thoughts on why 1 might be better than the other?

Reply:
nothing that matters. I tend to favor getters in case there is logic in them that i don't want to miss, but I know in this case there isn't. I'll default to directly accessing the object in future